Capture user input via internal API

bmuschko opened the pull request:
### Context

See https://github.com/gradle/gradle/issues/2837.
Branch build: https://builds.gradle.org/project.html?projectId=Gradle_Check&branch_Gradle_Check=bm%2Fcapture-user-input

**Overview**

- Internal API for capturing user input
- Capturing user input works in plain and rich console (if interactive)
- Capturing user input works with and without daemon
- User input can be requested for any answer
- User input can be requested for a restricted list of answers
- User input is sanitized
- Entering ctrl-d during user input returns `null` (though we could think about failing the build via `GradleException` or skipping the task with `StopExecutionException`)
- Currently unhandled: Detection of non-interactive console throws an exception (right now return `null` for user input). How exactly were you planning to handle that situation?

I'd like to hear your feedback as the build scan plugin will be first consumer of the functionality.

[![asciicast](https://asciinema.org/a/jnQ08PmmFX8CfMh7qzG4kD2re.png)](https://asciinema.org/a/jnQ08PmmFX8CfMh7qzG4kD2re)

breskeby commented:
that was me ☝️

bmuschko commented:
> 1. Do we plan to store answers in the future (I don't want to agree on a license over for every build again, though we could hint how to apply the license in the build script persistently)

IMHO that is somewhat out of scope for this change. I'd expect the plugin using the API to implement such functionality. We can't really anticipate _what_ the logic is going to do with the value after it was entered by the user. To conclude: not planned right now. In the long run we could think about the capability in Gradle core that stores value (however, we'd have to consider aspects like TTL, encryption of captured value etc). This sounds more like something we'd want to add with https://github.com/gradle/gradle/issues/2842.

> 2. Is checking for non interactive console enough to ensure this doesn't block for example CI builds?

The behavior at the moment: If you run in a non-interactive mode then the returned value from the API is going to be `null`. Even if you call the API in a non-interactive mode then it wouldn't block. We could provide some mechanism that a plugin can use to ask if the environment is interactive.

> 3. Should we probably provide a default mechanism?

That's not a bad idea. I was actually thinking about it as well but didn't want blow up the change. I think we can probably add something like that fairly easily. Will look into it.

> 4. Do we consider having a timeout on providing user input?

Was also wondering about it but couldn't really come up with a use case. In a non-interactive environment the user would not be asked. In an interactive environment the build would block on the question until the user enters `ctrl-c`/`ctrl-d` or enters an answer. Right now I can't think of reason why that wouldn't be sufficient.

breskeby commented:
>> 1. Do we plan to store answers in the future (I don't want to agree on a license over for every build again, though we could hint how to apply the license in the build script persistently)

>IMHO that is somewhat out of scope for this change. I'd expect the plugin using the API to implement such functionality. We can't really anticipate _what_ the logic is going to do with the value after it was entered by the user. To conclude: not planned right now. In the long run we could think about the capability in Gradle core that stores value (however, we'd have to consider aspects like TTL, encryption of captured value etc). This sounds more like something we'd want to add with https://github.com/gradle/gradle/issues/2842.

Sounds good to me and shouldn't block build scan plugin adoption for this feature. We can provide a hint how to make this persistence for now by accepting license in the plugin.

>> 2. Is checking for non interactive console enough to ensure this doesn't block for example CI builds?

>The behavior at the moment: If you run in a non-interactive mode then the returned value from the API > is going to be `null`. Even if you call the API in a non-interactive mode then it wouldn't block. We could provide some mechanism that a plugin can use to ask if the environment is interactive.

that's good enough from my point of view and not blocking build scan plugin adoption.

>> 3. Should we probably provide a default mechanism?

> That's not a bad idea. I was actually thinking about it as well but didn't want blow up the change. I think 
 we can probably add something like that fairly easily. Will look into it.

nice!

>> 4. Do we consider having a timeout on providing user input?

> Was also wondering about it but couldn't really come up with a use case. In a non-interactive environment the user would not be asked. In an interactive environment the build would block on the question until the user enters `ctrl-c`/`ctrl-d` or enters an answer. Right now I can't think of reason why that wouldn't be sufficient. 

I was more concerned by non interactive environments. It seems not necessary right now. 

From my point of view this PR looks good. would be great if @ldaley can have a look to I think

bmuschko commented:
>Sounds good to me and shouldn't block build scan plugin adoption for this feature. We can provide a hint how to make this persistence for now by accepting license in the plugin.

As an alternative to capturing user input from the console I could envision us just asking the user to create a property file in a specific location with a mandatory key/value pair. A message on the console would clearly indicate where the file should exist and how the contents should look like. If the property file or the key/value does not exist then fail the build if --scan was requested without the necessary configuration.

> I was more concerned by non interactive environments. It seems not necessary right now.

Certainly something we should make sure is working properly when you integrate the feature.

bmuschko commented:
@oehme Would you mind giving this a review as well?

etiennestuder commented:
There needs to be a way that is as easy as possible for the user. Creating a properties file is not. We _could_ provide the properties file approach as a secondary path, but definitely not as the first path of choice.

oehme commented:
Regarding the scope of this story: I'd propose we keep it simple and provide a way for the scan plugin to ask one specific question: "Do you accept the license? Yes/No". The scan plugin can build code around this that 

- looks up a property before asking the question
- stores the answer in gradle.properties so it doesn't have to ask again

We can enhance these capabilities later so that the user input API also takes over property lookup and storing answers.

bmuschko commented:
@oehme Thanks for your additional feedback. I addressed your comments. Can you have another look?

> I'd propose we keep it simple and provide a way for the scan plugin to ask one specific question.

I removed the default value functionality for now to keep the changes to a minimum.

> Instead there should be a very simple service above the domain layer that solves exactly the issue that the scan plugin needs to solve.

Good point, done.

oehme commented on subprojects/core/src/main/java/org/gradle/api/internal/tasks/userinput/BuildScanUserInputHandler.java:
```diff
@@ -0,0 +1,38 @@
+import java.util.List;
+
+@UsedByScanPlugin
+public class BuildScanUserInputHandler {
```
❌ This should be an interface.

bmuschko replied:
done

oehme commented on subprojects/core/src/main/java/org/gradle/api/internal/tasks/userinput/UserInputHandler.java:
```diff
@@ -0,0 +1,22 @@
+
+public interface UserInputHandler {
+
+    String getInput(InputRequest inputRequest);
```
⭕️ I would go as far as reducing this to `askYesNoQuestion`, because that's the only use case we have so far. That would reduce the complexity of this PR.

oehme commented on subprojects/logging/src/main/java/org/gradle/internal/logging/console/AbstractUserInputRenderer.java:
```diff
@@ -0,0 +1,87 @@
+    private void replayEvents() {
+        ListIterator<OutputEvent> iterator = eventQueue.listIterator();
+
+        while (iterator.hasNext()) {
```
⭕️ You could use the `Queue` interface for this.

bmuschko replied:
I believe for that to work, `OutputEvent` would have to implement `java.lang.Comparable`. Not going to do that for now.

oehme replied:
I'm not sure what you mean. You can use an ArrayDeque and use take() instead of get() and remove().

bmuschko replied:
Fair enough, using `ArrayDeque`.

bmuschko replied:
Had to revert that as some integration tests failed due to Java 5 compatibility.

pioterj commented:
Do we have an issue to make it public?

oehme commented:
Yes, see the attached epic.
